articles/node_lifecycle.md

```diff
@@ -0,0 +1,287 @@
+---
+layout: default
+title: Managed nodes
+abstract:
+  This article describes the concept of a node with a managed life cycle. It
+  aims to document some of the options for supporting managed-life cycle nodes
+  in ROS 2. It has been written with consideration for the existing design of
+  the ROS 2 C++ client library, and in particular the current design of
+  executors.
+author: '[Geoffrey Biggs](https://github.com/gbiggs)'
+published: true
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+A managed life cycle for nodes allows greater precision over the state of ROS
+system. It will allow roslaunch to have greater control over launching a ROS
+system, and in particular will allow roslaunch to ensure that all components
+have been instantiated correctly before it allows any component to begin
+executing its behaviour. It will also allow nodes to be restarted or replaced
+on-line.
+
+The most important concept of this document is that a managed node presents a
+well-known interface, executes according to a well-known life cycle state
+machine, and otherwise can be considered a black box. This allows freedom to
+the node developer on how they provide the managed life cycle functionality,
+while also ensuring that any tools created for managing nodes can work with any
+compliant node.
+
+
+## Life cycle
+
+![The proposed node life cycle state machine](/img/node_lifecycle/life_cycle_sm.png "The proposed node life cycle state machine")
+
+The behaviour of each state is as defined below.
+
+### Created state
+
+This is the life cycle state the node is in immediately after being
+instantiated. The node may transition to the `inactive` state by being
+initialised.
+
+Implementation note: In object-oriented languages, the onCreated method can be
+mapped to the class constructor.
+
+### Inactive state
+
+This state represents a node that is not currently performing any processing.
+While in this state, the node will not receive any execution time to read
+topics, perform processing of data, respond to service requests, etc. Any data
+that arrives on topics will not be read. Any service requests will not be
+answered.
+
+[To discuss: All of the above, and additionally should data/service requests be
```

Comment:
The goal to keep all services as much the same as possible is a good one. But the need I stated above still exists.

If all services can be called while the node is in the inactive state, then it is not really safe to reconfigure the node in this state. A service call made by another (active) node just before a configuration parameter changes might return a different result to one made just after the change, for example. The response time of service calls may also not be guaranteable.

On the other hand, if no services can be called while the node is inactive, then you have a jammed node because you can't call the service that activates it.

Off the top of my head, I can see two solutions to this problem. The first is hinted at in your reply: Control whether execution time is available for processing service requests or not. This would need to distinguish between functionality services and infrastructure services, so that execution time can be provided for infrastructure service requests. Whether or not the same executor provides time for both would need to be decided.

The second is to shift all the relevant infrastructure stuff to something that can service the requests, such as the executor.

It's worth noting here a couple of interesting things from the Orocos life cycle. The first is that Orocos allows for the source of execution time for handling service calls to be chosen from either the caller or the callee (the default). Obviously this is only possible when two components are running in the same process, which is the major use case of Orocos. It's an interesting concept, though. The other thing, which may or may not be true any more (I can't find it in the manual, but it was definitely there a few years ago) is that service calls are still responded to when the component is in the Stopped state.